Adds in Xenochimeras (Scree Whitelisted race)

Cameron653 opened the pull request:
**This is a New whitelisted race.**

**Proposed changes in this Pull Request:**
- Add in Xenochimera (Scree's race)
- Add in "chimera_revive" verb

**Notes**:
Added in Xenochimera race for Scree since I had a day of free time and it sounded like fun.

**CONS**:
- Completely unclonable.
- When hunger gets too low, they lose all ability to use most complex items until they eat, going into a feral mode.
- (Guns, traps, navbeacon, constructing, keycard reader, handcuffs, suit storage units, APC's, and disposal units are all unable to be used while in this state.)

**NEUTRAL**:
- Gets messages about being hungry when their hunger drops below a certain amount.
- Only one person can play this race at a time. Any more and bugs will happen. Not a problem since it was made for a single person.
- Gets its own custom egg_type
- Has a visible twitch when they're in feral mode.

**PROS**:
- Has a "revive" verb that is no faster than cloning. Requires nutriment in blood stream to revive if dead. Otherwise, can regenerate without nutriment required.

**REGENERATE VERB**:
- Based on current nutrition
- Fixes all ailments
- Minimum time to revive: 5 minutes
- Maximum time to revive: 30 minutes
- Has an hour in between uses.
- Cuts nutrition in half when used
- When using the regenerate verb, the user is completely knocked down, unable to move/defend theirself, completely helpless. Still able to talk/use /me's for RP.
- After regeneration is finished, gibs spew from the user and gives everyone in the immediate area text saying that X just suddenly exploded, much like the gib text.
- Requires nutriment injected into blood stream to revive from death.
- Drops the users items after regeneration.
- ~~Deals 10 damage after regeneration/revival~~ Removed due to "they are bruised" prompts.
- If reviving from death, gives 10 brain damage.
- Regeneration will give an ETA. (If it's 1000 seconds, it'll say "Approximately 1000 seconds until revival."
- User won't suffocate/bleed while rejuvenating
- If the user dies while regenerating, they will require nutriment in bloodstream to revive.
- Deals 60 clone damage when rejuv'd, which is then slowly healed over time. 1 dmg per min heal.
- Gives a hatch "verb" that lets the user revive/rejuvenate whenever they want, provided they have nutriment in them (If they were previously dead). After hatch is used, they will need to rejuvinate again (In another hour after the initial hatch)

**EXAMPLES**:
0 nutrition, 0 would be used, timer would be thirty minutes.
100 nutrition, 50 would be used, timer would be 1366 seconds (Aprox 23 minutes)
200 nutrition, 100 would be used, timer would be 1150 seconds (Aprox 19 minutes)
400 nutrition, 200 would be used, timer would be 933 seconds (Aprox 15.5 minutes) Starting Nutrition
500 nutriton, 250 would be used, timer would be 871 seconds (Aprox 14.5 minutes)
2000 nutrition, 1000 would be used, timer would be 618 seconds (Aprox 10 minutes)
20000 nutrition, 10000 would be used, timer would be 506 seconds (Aprox 8.5 minutes) (The lowest)

Notes:
Was having a conversation about Prometheans and their health regeneration, and it ended up in a conversation about alternatives to health regenerating abilities, which led to me creating this race for Scree. It has one major trait (Being able to rejuvenate theirself) but also has cons that balance it out (Being unclonable, and being able to use any actual major items while starving), and also has features that lead to better RP (Race becomes feral if starving.). Not only that, but another thing that makes the rejuvenate not that great is its nutrition great, and the flat out amount of time you'll be vulnerable for while using it. I attempted to make it as balanced as possible.

**MISC THINGS ADDED/FIXED ALONG THE WAY**:
- Added the old cobra hood sprite.
- Fixed a problem with digestion nutrition gain when the prey weighed < 137
- is_whitelisted removed from all species, allowing whitelist to be enabled.

Screemonster commented:
I was mostly joking about this but he actually did it _the absolute madman_

I suspect this'll be more of a precursor to some of the !!FUN!! that's being talked about with the planned traits thing.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_vr.dm:
```diff
@@ -0,0 +1,84 @@
+
+	//primitive_form = "Farwa"
+
+	spawn_flags = SPECIES_CAN_JOIN || SPECIES_IS_WHITELISTED
```
Single pipe, not double. Also we have whitelisting turned off so this needs to be RESTRICTED not WHITELISTED. The former means only admins can join as it.

Arokha commented:
There is a file somewhere that has the type paths for /mob/living/carbon/human/tajaran and such. This needs an entry in that file so it shows up in the debug spawn menu. All humanoids are actually the same path but the extra little stub has an overridden New that sets their species.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_abilities_vr.dm:
```diff
@@ -0,0 +1,44 @@
+	set name = "Chimera Regenerate"
+	set category = "Abilities"
+
+	if(world.time < last_special)
```
Or if regenerating? Wouldn't they be able to start two regens after 50 seconds? Maybe you should make this a clickcooldown thing. Look at belly_vr relayresist.

SpadesNeil commented:
Regarding the 'revive' verb, how does it work? In my experience, to resurrect, Scree needed enough biomatter in him to do the trick. In this case, would straight up injecting the deceased corpse with nutriment do the trick? It'd have to be a lot of nutriment.

Arokha commented:
In terms of the rig_pieces and suit cycler stuff, can you remove that? He has his own fluff spacesuit already. I'm gonna say if we do future species, they need to bring their own equipment. Adding to those lists is too snowflakey for me. They can just bring a modkit.

Screemonster commented:
>Regarding the 'revive' verb, how does it work? In my experience, to resurrect, Scree needed enough biomatter in him to do the trick. In this case, would straight up injecting the deceased corpse with nutriment do the trick? It'd have to be a lot of nutriment.

In this case, he'd be essentially cannibalising his own body and consuming an absolute fuckton of nutriment (literally half what he has, so several hundred in most cases). It's the fluff thing of what he's been doing whenever he goes home with a hole blown through his torso by security _again._

The nutrient bath thing is a fluff thing for Central to get it to go faster since the time taken to regenerate scales up massively when he's not a fatty with plenty of nutrient reserves to spare.

Screemonster commented:
Oh! That reminds me.
@Cameron653 - Scree's regen is _not_ good at dealing with brain damage, so just cap the brainloss at 50 or something. If he's regenerating from "dead" then _give_ him some as well.

SpadesNeil commented:
Yea there needs to be some kind of drawback to death. I don't want anyone to have what is basically immortality. Assume we might let normal players use this. I don't like adding admin abuse races.

Cameron653 commented:
I had a nice reply setup but github ate it :v 

>There is a file somewhere that has the type paths for /mob/living/carbon/human/tajaran and such. This needs an entry in that file so it shows up in the debug spawn menu. All humanoids are actually the same path but the extra little stub has an overridden New that sets their species.

done

>Or if regenerating? Wouldn't they be able to start two regens after 50 seconds? Maybe you should make this a clickcooldown thing. Look at belly_vr relayresist.

Good catch. Fixed this and made it so all races have the "reviving" variable, just in case an admin event gives a race other than the chimera the revive command for the duration of it. (Like giving a shadowperson it)

>Regarding the 'revive' verb, how does it work? In my experience, to resurrect, Scree needed enough biomatter in him to do the trick. In this case, would straight up injecting the deceased corpse with nutriment do the trick? It'd have to be a lot of nutriment.

It takes the nutrition they had at the time of death, slices it in half, and uses that half to do the regeneration. Time to revive varies by how much nutrition they have. Has a curve made to prevent abuse & stocking up on nutrition not worth it. Injecting nutrition while they are alive would allow them to revive faster, but injecting it after they died (Now) goes and allows them to revive from death. Otherwise, they can't revive without nutrition inside of them.

>In terms of the rig_pieces and suit cycler stuff, can you remove that? He has his own fluff spacesuit already. I'm gonna say if we do future species, they need to bring their own equipment. Adding to those lists is too snowflakey for me. They can just bring a modkit.

Done. Removed them from rig peices.

>Oh! That reminds me.
@Cameron653 - Scree's regen is not good at dealing with brain damage, so just cap the brainloss at 50 or something. If he's regenerating from "dead" then give him some as well.

Fixed and gave it

>Yea there needs to be some kind of serious drawback to death. I don't want anyone to have what is basically immortality. Assume we might let normal players use this. I don't like adding admin abuse races.
You shouldn't be able to resurrect yourself. I dislike that idea a lot. You need outside help.

Tweaked it so now if a Chimera dies and tries to revive they have to have nutrition in their blood stream. If they don't, they can't revive. If they're alive when they try to do this they don't need to do so.

Cameron653 commented:
Alright, everything suggested has been fixed/added/adjusted.

Cameron653 commented:
Apparantly something broke
Don't merge yet

Cameron653 commented:
Problem fixed.

ALSO, SIDE NOTE:
WHITELIST MUST BE ENABLED.
If whitelist isn't enabled, anyone will be able to pick Xenochimera. Just make sure that Scree is the only one able to play Xenochimera.

Arokha commented:
I ain't done lookin' at it :v

Screemonster commented:
Question.
If a race is whitelisted/restricted, will the option still show up to taunt people who don't have it?

Arokha commented:
> If a race is whitelisted/restricted, will the option still show up to taunt people who don't have it?

Something I'd like to test.

Cameron653 commented:
>Question.
If a race is whitelisted/restricted, will the option still show up to taunt people who don't have it?

If it's true to every other codebases whitelist system, it'd show up as a playable character and then  tell you "No" if you tried to join in as it.

SpadesNeil commented:
Except our whitelist for species have been FUBAR lately.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_vr.dm:
```diff
@@ -0,0 +1,84 @@
+	unarmed_types = list(/datum/unarmed_attack/stomp, /datum/unarmed_attack/kick, /datum/unarmed_attack/punch, /datum/unarmed_attack/bite)
+	num_alternate_languages = 2
+	secondary_langs = list("Sol Common")
+	color_mult = 1
```
Tajaran sprites are meant for color_add, so you should remove this. Makes them really hard to color.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_vr.dm:
```diff
@@ -0,0 +1,84 @@
+	name = "Xenochimera"
+	name_plural = "Xenochimeras"
+	icobase = 'icons/mob/human_races/r_tajaran.dmi' //Tajara basis, as it has working skin color.
+	deform = 'icons/mob/human_races/r_def_tajaran.dmi'
```
Needs some tail-spritey-bits up in here. Look at Tajaran species def.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_vr.dm:
```diff
@@ -0,0 +1,84 @@
+		if(M != H)
+			if(prob(0.5))//1 in 200 chance to tell them that person looks like food. This is so irregular so it doesn't pop up 24/7 during ERP.
+				H << "<span class='danger'> You feel a stabbing pain in your gut, causing you to twitch in pain.. It would be extremely wise to find some type of food... In fact, [M] looks extremely appetizing...</span>"
+				H.emote("twitch")
```
This comes up a bit when you're dying, though. Like you're in crit, and it's like "You can't emote right now!"

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_abilities_vr.dm:
```diff
@@ -0,0 +1,72 @@
+				C << "You're dead and have no liquid nutriment to use for the regeneration process."
+				return
+
+		else if(C.stat == CONSCIOUS || C.stat == UNCONSCIOUS) //If they're alive
```
C.stat != DEAD or C.stat < DEAD if you're worried we'll add something above DEAD later.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_abilities_vr.dm:
```diff
@@ -0,0 +1,72 @@
+				C.weakened = 10000 //Since it takes 1 tick to lose one weaken. Due to prior rounding errors, you'd sometimes unweaken before regenning. This fixes that.
+				C.species.reviving = 1
+				C.canmove = 0 //Make them unable to move. In case they somehow get up before the delay.
+				C << "You begin to reconstruct your form. You will not be able to move during this time."
```
Maybe it could tell you how long it'll take? It's kinda frustrating to have zero idea when it'll be done so you can't go get a sandwich or anything.

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_abilities_vr.dm:
```diff
@@ -0,0 +1,72 @@
+			C << "You begin to reconstruct your form. You will not be able to move during this time."
+
+			spawn(time SECONDS) //Not going to do another "if stat = dead" check because that'd make me have to do more lines than needed.
+				C << "<span class='notice'>We have reconstituted our form.</span>"
```
This spawn needs some checks when it comes to term.

- Check to see if C still exists, otherwise it will runtime if it's been deleted or gibbed/husked meanwhile.
- Check to see if C died while reconstructing, since if that's the case, you need to handle it some other way, otherwise as long as you click the button before you die, you're invincible.

Arokha commented:
How are you preventing them from getting ghosted when dead?

Arokha commented:
If you are bleeding, there is a VERY high chance of bleeding to death in my testing. I lopped off a leg and did stasis, and died in about a minute. Just an FYI, really, to make sure that's what you intended. Maybe it should clean up lost limbs or something at first? I dunno.

Cameron653 commented:
>How are you preventing them from getting ghosted when dead?

With the getting ghosted, even if you ghost you can then return back to your body and do the proc (If you have nutriment in you)

>If you are bleeding, there is a VERY high chance of bleeding to death in my testing. I lopped off a leg and did stasis, and died in about a minute. Just an FYI, really, to make sure that's what you intended. Maybe it should clean up lost limbs or something at first? I dunno.

I indeed made it so that if you're undergoing regeneration and _then_ die, you don't need nutriment in you (I could change this if requested). If you die and then try to do it, you need nutriment in you. After you revive, your  blood and everything else is fixed back up, just like new.

Arokha commented:
> I indeed made it so that if you're undergoing regeneration and then die, you don't need nutriment in you

But that means that you can't ever die. :v Like you can click this button whenever and just be fine. S'what you want?

Arokha commented on code/modules/mob/living/carbon/human/species/station/station_special_abilities_vr.dm:
```diff
@@ -0,0 +1,72 @@
+					new /obj/effect/gibspawner/human/scree(T)
+					var/braindamage = C.brainloss/2 //If you have 100 brainloss, it gives you 50.
+					C.species.reviving = 0
+					C.revive() // I did have special snowflake code, but this is easier.
```
This seems to do 'some stuff' that is unexpected. Like if it replaces limbs that were lost that had items on them or something then the items are attached to you forever. Like I lost a leg, and I could then not remove my shoes after it finished. I was also walking slow even though there was no reason to be.

Cameron653 replied:
Hmm, good catch. I'll code it to drop all the items that the user was previously wearing on the floor, since they _are_ exploding.

Cameron653 replied:
This is now done. Items drop upon regenerating.

Cameron653 commented:
>But that means that you can't ever die. :v Like you can click this button whenever and just be fine. S'what you want?

Thinking about it, I probably _should_ change it so that if you die while regenerating that you still need nutrition. Would be easy enough to do.

Cameron653 commented:
Everything above is fixed/added in.

SpadesNeil commented:
I won't allow this to be added unless outside help is required to revive a Xenochimera.

Cameron653 commented:
>I won't allow this to be added unless outside help is required to revive a Xenochimera.

Err... I already implemented that 9 commits ago :v 
Chimeras that die need someone to inject nutriment after death into them to revive.

Cameron653 commented:
Also, need some input:

Should it be that if you begin the rejuvenation and then die, you need nutriment in you (as it currently coded in as)

Or should it be that if you begin the rejuvenation and then die, you revive anyways, without the need of nutriment?

SpadesNeil commented:
>Err... I already implemented that 9 commits ago :v
>Chimeras that die need someone to inject nutriment after death into them to revive.

shut up I'm sleepy

Cameron653 commented:
Everything appears to be fixed/done/working. Does anyone want to give any more input on this? I went ahead and testing it and it seems that all the bugs have been ironed out.

Screemonster commented:
I can't think of any other racial traits that'd be appropriate. Certainly not giving 'em a lingsting, that'd be seriously broken unless it had some serious drawbacks to make it impractical to use in an actual fight. (in which case we can just RP that shit like I already do)

He's scuttly but probably not to the extent of tajara turbospeed. No real resistances/vulnerabilities to speak of. Has claws. I don't actually know what things are options.
Hell, all these things could probably be left until the traits thing is put in.

SpadesNeil commented:
Include Xenochimera on the wiki and we're done.

Screemonster commented:
https://wiki.vore-station.net/Xenochimera

Although, I did come up with a possible implementation of the cooldown that might be better than a hard "has it been long enough?" timer.

Basically, use cloneloss.

1: upon standing up after regeneration/hibernation, set cloneloss to 60. (helpfully represents the fact that regenerating is _fucking crippling_ and leaves them in a severely weakened state)
2: let 'em _regenerate_ cloneloss at a rate of roughly 1/minute at a cost of yet more nutrition (balance question? Cloneloss is normally rare and really hard to treat, it'd be fairly unique for a race to be able to fix their own genetic damage, but still waaaay slower than actual medical treatment)
3: for every point of cloneloss when they hit the regen button, add _two_ minutes onto the timer for how long their ass is on the ground. So someone hitting the button "early" will still hibernate, but at a _huge_ time penalty compared to staying on their feet and waiting it out.

Would be helpful if there could be some downside to using the cryotubes so they can't just use that to recover. Possibly make cryoxadone/clonexadone ineffective or provide some other penalty for using it, or have the low temperatures required for those chems to work just weaken them for a stupendous period of time.

Cameron653 commented:
Good idea, and it sounds easy enough to do. Give me a bit and I should have it up and working.

Cameron653 commented:
Err... I had the cloneloss setup and everything, but as it turns out.... If you have the NO_SCAN flag, your clone loss is automatically set to 0. So it doesn't work. Replacing it with toxinloss wouldn't work, either, as you'd be puking up everywhere and that would be awful for RP.

Screemonster commented:
Welp. Leave it as we had it, was just an idea.

Cameron653 commented:
Updates to allow people to spawn in as all races (with the exception of xenochimera) if/when the whitelist is enabled.

Cameron653 commented:
Doing one final edit and this'll be ready for merge. Going to remove the 10 brute damage as it causes "They have bruises all over their X", which isn't good.

Cameron653 commented:
Alright, it should be ready for merge. Everything appears to be working.

SpadesNeil commented:
>Otherwise, can regenerate without nutriment required.

This should be removed. No nutriment, no revive.

Cameron653 commented:
Regenerate, not revive.

Regenerating heals damage they take, but not If they're dead. Them having
no nutrition while doing it isn't a big deal, as they'll be unable to do
anything for a long while.

Reviving heals them back from death and can only be preformed while they
have nutriment injected into them.

Also, I'm not at my house currently, so I can't remove the "gets 60 clone
damage and heals said clone damage over time" bulletpoint. This was a
feature which was planned and scrapped.

On Wed, Jun 22, 2016, 2:06 PM Spades <notifications@github.com> wrote:

> Otherwise, can regenerate without nutriment required.
>
> This should be removed. No nutriment, no revive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/VOREStation/VOREStation/pull/350#issuecomment-227828488>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/APOt87zDS83wR0B-uzNMBU1Z2jl6ClDRks5qOXmwgaJpZM4IzfoM>
> .
>

SpadesNeil commented:
Okay this is good to go then.
